src/main/java/com/google/api/codegen/config/CollectionPattern.java

```diff
@@ -0,0 +1,135 @@
+/* Copyright 2016 Google Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.api.codegen.config;
+
+import com.google.api.codegen.Inflector;
+import com.google.api.tools.framework.aspects.http.model.HttpAttribute.FieldSegment;
+import com.google.api.tools.framework.aspects.http.model.HttpAttribute.LiteralSegment;
+import com.google.api.tools.framework.aspects.http.model.HttpAttribute.PathSegment;
+import com.google.api.tools.framework.aspects.http.model.HttpAttribute.WildcardSegment;
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Class representing a CollectionPattern that will be output to the collection config section
+ */
+public class CollectionPattern {
+
+  public static CollectionPattern create(FieldSegment fieldSegment) {
+    if (!isValidCollectionPattern(fieldSegment)) {
+      throw new IllegalArgumentException("Field segment is not a valid collection pattern.");
+    }
+    return new CollectionPattern(fieldSegment);
+  }
+
+  private final FieldSegment fieldSegment;
+  private final List<PathSegment> templatizedSubpath;
+  private final String simpleName;
+
+  private CollectionPattern(FieldSegment fieldSegment) {
+    this.fieldSegment = fieldSegment;
+
+    ImmutableList.Builder<PathSegment> builder = ImmutableList.<PathSegment>builder();
+    PathSegment lastSegment = null;
+    String simpleName = null;
+    for (PathSegment pathSegment : fieldSegment.getSubPath()) {
+      if (pathSegment instanceof WildcardSegment) {
+        String name = "unknown";
+        String suffix = "";
+        if (lastSegment != null && lastSegment instanceof LiteralSegment) {
+          name = Inflector.singularize(lastSegment.syntax());
+        }
+        if (((WildcardSegment) pathSegment).isUnbounded()) {
+          name += "_path";
+          suffix = "=**";
+        }
+        builder.add(new LiteralSegment(String.format("{%s%s}", name, suffix)));
```

Comment:
If you're just concatenating name and suffix here, but you're adding things to name earlier, you might as well just have a single string that you keep adding stuff to, and pass that one string into the format here.